xtensa-build-zephyr.py: pass SOF firmware file version string to rimage

mengdonglin opened the pull request:
This is to fix https://github.com/thesofproject/sof/issues/5775. It depends on rimage PR https://github.com/thesofproject/rimage/pull/95

Extract SOF firmware file version string (x.y or x.y.z) from latst git tag
by 'git describe --tags' and python regex, and pass it to rimage in
signing phase. So rimage can write the version info to the firmware
manifest headers for cAVS platforms.

Signed-off-by: mengdonglin <mengdong.lin@intel.com>

aiChaoSONG commented on scripts/xtensa-build-zephyr.py:
```diff
@@ -447,6 +448,10 @@ def build_platforms():
 			signing_key = default_rimage_key
 		sign_cmd += ["--tool-data", str(rimage_config), "--", "-k", str(signing_key)]
 
+		sof_tag = subprocess.check_output(["git", "describe", "--tags"]).strip().decode()
```
This is really an independent feature, let's make it a function, for example, `extract_sof_version_from_git()` or simplely `extract_sof_version`, let's hear more from @marc-hb

mengdonglin replied:
@aiChaoSONG Okay. Thank you!  I updated the PR to define a function get_sof_version() to get SOF FW file version string.

marc-hb commented on scripts/xtensa-build-zephyr.py:
```diff
@@ -323,6 +324,14 @@ def west_init_update():
+	by 'git describe --tags' and regex match
+	"""
+	sof_tag = subprocess.check_output(["git", "describe", "--tags"]).strip().decode()
+	sof_version = re.search('([0-9]+)[.]([0-9]+)([.]([0-9]+))?', sof_tag)
```
Please don't copy/paste/diverge the code already in `scripts/cmake/version.cmake`. It's more complex than this and already different from this. Also, `git describe` can take time.

Instead parse the output of `scripts/cmake/version.cmake` like this (tested):
```python
defines = {}
with open("zephyrproject/build-$platform/zephyr/include/generated/sof_versions.h") as hfile:
    for hline in hfile:
        words = hline.split()
        if words[0] == '#define':
            defines[words[1]] = words[2]

print(defines)

{'SOF_MAJOR': '3', 'SOF_MINOR': '2', 'SOF_MICRO': '1', 'SOF_TAG': ...
```

mengdonglin replied:
@marc-hb I updated the PR to get version from sof_version.h as you suggested, and we can always pass a 3-field version string to rimage.
@aiChaoSONG @lgirdwood We no longer need the rimage PR https://github.com/thesofproject/rimage/pull/95 so I closed it.

marc-hb commented on scripts/xtensa-build-zephyr.py:
```diff
@@ -447,6 +463,11 @@ def build_platforms():
 		sign_cmd += ["--tool-data", str(rimage_config), "--", "-k", str(signing_key)]
 
+		global sof_version
+		if sof_version == "0.0.0":
```
Can you move this test inside `get_sof_version()` ? Making the cache a private implementation detail.

```python
def get_sof_version(abs_build_dir):
    if sof_version:
          return sof_version

    with open(...

       sof_version = ...

    return sof_version
```

mengdonglin replied:
Okay, moving this test inside get_sof_version() is better. Thanks for your suggestions @marc-hb I updated the PR to address all your review comments.

marc-hb commented on scripts/xtensa-build-zephyr.py:
```diff
@@ -23,6 +23,9 @@
 default_rimage_key = pathlib.Path("modules", "audio", "sof", "keys", "otc_private_key.pem")
 
+# Initial value of SOF firmware file version
+sof_version = "0.0.0"
```
```suggestion
sof_version = None
```

mengdonglin replied:
Updated. Set sof_version is initialized to None now.

marc-hb commented on scripts/xtensa-build-zephyr.py:
```diff
@@ -323,6 +326,19 @@ def west_init_update():
 
+def get_sof_version(abs_build_dir):
+	"""[summary] Generate SOF FW file version string major.minor.micro from sof_version.h
+	When builing multiple platforms from the same SOF commit, all platforms share the same version.
```
```suggestion
	When building multiple platforms from the same SOF commit, all platforms share the same version.
```

mengdonglin replied:
Function description has been updated and typo is fixed

aiChaoSONG commented on scripts/xtensa-build-zephyr.py:
```diff
@@ -323,6 +326,25 @@ def west_init_update():
+	So for the 1st platform, generate the version string from sof_version.h and later plaforms will
+	reuse it.
+	"""
+	global sof_version
```
The feature can be done by:
- use a global sof_version, and modify it with `get_sof_version`. in signing, use the global one directly.
- don't use a global sof_version, and in signing, use the one returned from `get_sof_version`

Seems you mixed use above two in the code, I prefer the second one without the global version. so the code of `get_sof_version` can be simplified as:

```

versions = {}
with open(pathlib.Path(abs_build_dir, "zephyr/include/generated/sof_versions.h")) as hfile:
		for hline in hfile:
			words = hline.split()
			if words[0] == '#define':
				versions[words[1]] = words[2]
return versions['SOF_MAJOR'] + '.' + versions['SOF_MINOR'] + '.' + versions['SOF_MICRO']
```

marc-hb replied:
I agree with @aiChaoSONG that there is a problem now: the `sof_version` global is set twice, BOTH inside and outside the function. That's not good.

However I prefer the other option: keep all read/write accesses to the global `sof_version` and its "maintenance" _private_ and isolated inside the function.

marc-hb replied:
Oh wait: `sof_version` is local, shadowing the global. That's not good either; shadowing == bad.
```
pylint scripts/xtensa-build-zephyr.py| sort -t: -k2,3 -n

scripts/xtensa-build-zephyr.py:472:2: W0621: Redefining name 'sof_version' from outer scope (line 27) (redefined-outer-name)

```

https://pylint.pycqa.org/en/latest/user_guide/messages/warning/redefined-outer-name.html

Same solution.

marc-hb commented on scripts/xtensa-build-zephyr.py:
```diff
@@ -23,6 +23,9 @@
 west_top = pathlib.Path(SOF_TOP, "zephyrproject")
 default_rimage_key = pathlib.Path("modules", "audio", "sof", "keys", "otc_private_key.pem")
 
+# Initial value of SOF firmware file version
```
Nit: the comment does not seem to add much. Self-describing code FTW.

mengdonglin replied:
Will remove

marc-hb commented on scripts/xtensa-build-zephyr.py:
```diff
@@ -323,6 +326,25 @@ def west_init_update():
+		return sof_version
+
+	versions = {}
+	with open(pathlib.Path(abs_build_dir, "zephyr/include/generated/sof_versions.h")) as hfile:
```
New line

```suggestion
	with open(pathlib.Path(abs_build_dir,
	          "zephyr/include/generated/sof_versions.h")) as hfile:
```

Try `pylint` or `black` (warning: make a git commit or backup before using `black`)

marc-hb commented on scripts/xtensa-build-zephyr.py:
```diff
@@ -323,6 +326,25 @@ def west_init_update():
+			words = hline.split()
+			if words[0] == '#define':
+				versions[words[1]] = words[2]
+	sof_version = versions['SOF_MAJOR'] + '.' + versions['SOF_MINOR'] + '.' + versions['SOF_MICRO']
```
Newline 
```suggestion
	sof_version = versions['SOF_MAJOR'] + '.' + versions['SOF_MINOR'] +
	               '.' + versions['SOF_MICRO']
```
Again let `pylint` or `black` tell you the "right" way to format this.

mengdonglin replied:
Thanks for the tips @marc-hb  Will fix all the identification issues.

marc-hb commented on scripts/xtensa-build-zephyr.py:
```diff
@@ -447,6 +469,9 @@ def build_platforms():
 		sign_cmd += ["--tool-data", str(rimage_config), "--", "-k", str(signing_key)]
 
+		sof_version = get_sof_version(abs_build_dir)
+		sign_cmd += ["-f", sof_version]
```
```suggestion
		sign_cmd += ["-f", get_sof_version(abs_build_dir)]
```

mengdonglin commented:
@marc-hb No, I don't want to restart the review cycle.  I updated the PR: keep `sof_version` global, removed unnecessary comments for it, and separated the 2 long lines. I ran pylint to give a check. Could you check if it looks better now?

aiChaoSONG commented:
> However I prefer the other option: keep all read/write accesses to the global sof_version and its "maintenance" private and isolated inside the function.

A closure is a good fit for this.
```
def sof_version_getter():
    sof_version = None
    
    def get_sof_version(abs_path):
        nonlocal sof_version
        if sof_version:
            return sof_version
        read_and_parse sof_version_from_file
        return sof_version

    return get_sof_version

sign_cmd += ["-f", sof_version_getter()(abs_build_dir)]

```

mengdonglin commented:
Closure can make the cache local for the outer function and looks simpler than class.  What do you think @marc-hb ?

marc-hb reviewed (approved):
A closure sounds cool but I don't think the following code optimizes anything because the same `sof_version` is shared across all function calls (try: `print(id(sof_version))`. So `sof_version` is reset to `None` every time `sof_version_getter()` is called. It would work if the `get_sof_version` function returned by `sof_version_getter()` were remembered in a... `global`! But then we have more complicated code and still a global :-)

This is just a simple build script, I think a global `sof_version` is fine.

```python
def sof_version_getter():
    sof_version = None
    
    def get_sof_version(abs_path):
        nonlocal sof_version
        if sof_version:
            return sof_version
        read_and_parse sof_version_from_file
        return sof_version

    return get_sof_version

sign_cmd += ["-f", sof_version_getter()(abs_build_dir)]
```

aiChaoSONG commented:
> So sof_version is reset to None every time sof_version_getter() is called.

@marc-hb  You are right, to make the code work, the call to `sof_version_getter()` should be moved out of the for loop.
